File: backend/app/models/task.py
from __future__ import annotations

from pydantic import BaseModel, Field


class TaskNode(BaseModel):
    id: str
    title: str
    description: str
    estimated_hours: float = Field(ge=0)
    dependencies: list[str] = Field(default_factory=list)
    priority: str = Field(default="medium", pattern="^(low|medium|high|critical)$")
    children: list[TaskNode] = Field(default_factory=list)


class TeamMember(BaseModel):
    name: str
    hours_per_week: float = Field(gt=0)


class DecomposeRequest(BaseModel):
    prompt: str = Field(min_length=1, description="Natural language feature request")
    repo_context: str | None = Field(
        default=None,
        description="GitHub/GitLab branch URL or directory file listing",
    )
    team_members: list[TeamMember] = Field(default_factory=list)
    hours_per_week_default: float = Field(
        default=40, gt=0, description="Default hours per week per member"
    )
    sprint_days: int = Field(default=10, gt=0, description="Number of days per sprint")
    buffer_percent: float = Field(
        default=0.2, ge=0, le=1, description="Safety buffer percentage"
    )


class DecomposeResponse(BaseModel):
    tasks: list[TaskNode]
    total_estimated_hours: float
    sprint_capacity: float
    sprints_needed: int


class ScheduledTask(BaseModel):
    id: str
    title: str
    description: str
    estimated_hours: float
    dependencies: list[str]
    priority: str
    start_day: int
    end_day: int
    sprint: int


class SprintPlan(BaseModel):
    sprint_number: int
    tasks: list[ScheduledTask]
    total_hours: float
    capacity: float


class ScheduleResponse(BaseModel):
    sprints: list[SprintPlan]
    total_hours: float
    total_sprints: int
    capacity_per_sprint: float

File: frontend/src/types/task.ts
export interface TaskNode {
  id: string;
  title: string;
  description: string;
  estimated_hours: number;
  dependencies: string[];
  priority: "low" | "medium" | "high" | "critical";
  children: TaskNode[];
}

export interface TeamMember {
  name: string;
  hours_per_week: number;
}

export type DecomposeMode = "full_project" | "ai_ds_experiment";

export interface DecomposeRequest {
  prompt: string;
  mode: DecomposeMode;
  repo_context?: string | null;
  team_members: TeamMember[];
  hours_per_week_default: number;
  sprint_days: number;
  buffer_percent: number;
}

export interface ExpandTaskRequest {
  task: TaskNode;
  user_context?: string | null;
  mode: DecomposeMode;
}

export interface ExpandTaskResponse {
  parent_id: string;
  subtasks: TaskNode[];
}

export interface DecomposeResponse {
  tasks: TaskNode[];
  total_estimated_hours: number;
  sprint_capacity: number;
  sprints_needed: number;
}

export interface ScheduledTask {
  id: string;
  title: string;
  description: string;
  estimated_hours: number;
  dependencies: string[];
  priority: string;
  start_day: number;
  end_day: number;
  sprint: number;
}

export interface SprintPlan {
  sprint_number: number;
  tasks: ScheduledTask[];
  total_hours: number;
  capacity: number;
}

export interface ScheduleResponse {
  sprints: SprintPlan[];
  total_hours: number;
  total_sprints: number;
  capacity_per_sprint: number;
}

export interface ValidationResult {
  valid: boolean;
  dependency_errors: string[];
  circular_dependency_errors: string[];
  total_estimated_hours: number;
}
